packages/frontend/src/index.js

```diff
@@ -20,6 +20,10 @@ const history = createBrowserHistory();
 
 export const store = createStore(createRootReducer(history), createMiddleware(history));
 
+store.injectReducer = () => {
+    store.replaceReducer(createRootReducer(history));
```

Comment:
This feels a little heavy for just to key part of the redux state by account ID. Couple questions:
1. If I change accounts and then change back, does all the state for the original account have to be re-fetched?
2. What happens if we don't call `injectReducer`? Does the account-specific state get updated?
3. Could we get around needing to do this if we had top-level redux state for the currently active account?

Reply:
Ad 1. The idea is to keep the state, so the data will be instantly visible. But it's not possible for now, as we are purging `account` reducer on the account switch.
Ad 2. `injectReducer` is used only in the case of adding or recovering an account when we need to add a new reducer because it doesn't exist in the current state. Reducers for existing accounts are created right away based on the account list in local storage, but if we want to create a new reducer we need to dynamically add it using `replaceReducer`. If we add/recover a new account and don't add a new reducer it will not exist, no data will be stored.
Ad 3. Not sure what you mean :) could you please clarify?

Reply:
I see, so `saveAccount` is only used for adding a new account (either creation or recovery) to LocalStorage?

Looking at the `replaceReducer` documentation I see the outlined use cases are code-splitting and HMR - neither of which are conventional production runtime use cases like this usage is. It seems excessive to replace the entire root reducer just for adding one new account. If the intention here is to populate state for a new account ID key in `byAccountId`, then why do we need to replace the entire root reducer? What are the performance and UI implications of doing this?

If this is the way we need to do it then I think this needs to be feature flagged - with my limited understanding of `replaceReducer` I think I'm missing something crucial about how this new structure is supposed to work. It would be better to iterate on it if we're not yet sure what the final implementation should look like (i.e. `TODO: figure out better way to inject reducer`).

re: my comment on the top-level reducer, I am picturing this conceptually as a map of account IDs in the redux state whose value is some `AccountState` type (which, going from the diagram in the ticket, includes `tokenBalancesNFT`, `tokenBalancesFT`, etc.). Since you can only have one active account, the active account is tracked by another key at the top level of the state. So your state for a single account would be something like:
```
{
  shared: {
    fiatValues: { usd: 10.50 },
  },
  activeAccountId: 'spaghetti.near',
  byAccountId: {
    'spaghetti.near': {
      tokenBalancesNFT: { ... },
      ...
    },
  },
}
```

Then if I add a new account, it becomes:
```
{
  shared: {
    fiatValues: { usd: 10.50 },
  },
  activeAccountId: 'lasagna.near',
  byAccountId: {
    'spaghetti.near': {
      tokenBalancesNFT: { ... },
      ...
    },
    'lasagna.near': {
      tokenBalancesNFT: {},
      ...
    },
  },
}
```

So I've only added a key (with initial value) to the `byAccountId` map and updated `activeAccountId` to the new account. Other state (`shared`) is not impacted by this because it's outside the scope of what's changed by adding a new account. In my naive model here, it seems like I should be able to dispatch an `addAccount` action that:
- adds the new account ID
- initializes the reducer state under this new account ID key in `byAccountIds` to an empty state
- fires off requests to populate this new state with values from on chain + contract helper (only if recovering, new accounts probably wouldn't have any data on chain that's not already in context)

But in this implementation we're reinitializing and replacing the entire root reducer, which would include `shared`, even though non-account specific data hasn't changed. Given my understanding this seems excessive because we know exactly what account state needed to change when we added a new account - I don't see what makes this state different than any other in the store.

I see that `injectReducer` is mentioned explicitly in the diagram on the ticket, so I'm assuming that this is the plan - I'm just having trouble reconciling my understanding with this implementation. Maybe @MaximusHaximus can weigh in here with what the intention behind replacing the root reducer is.

Reply:
Thank you andy for checking these changes in detail. It's always very valuable, especially with such important changes.

> I see, so saveAccount is only used for adding a new account (either creation or recovery) to LocalStorage?

`saveAccount` is the last step of both creating and recovering an account, it's adding the account to the local storage and to Wallet `accounts` array. 

Using `saceAccount` for injecting reducer is a temporary solution (that's why the comment) because we don't want to use the `store` in the wallet.js (it's causing circular reference problems), but I'm using it there because with this PR I want to propose a concept of splitting the state tree. The next PR will be about moving `injectReducer` away from `saveAccount`, and preparing a more reasonable solution. Ideally, I would like to avoid adding `injectReducer` in a few places (for every way that we create or recover an account), I have a few concepts on how to do it - generally, I would like to have a thunk that will be dispatched as the last step of creating/recovering account (for every method) so we will be able to use it for every functionality that is common for every create/recover account method - in this case, it will be `injectReducer`, but it might be handy for future changes as well.

Btw I would love to organize the code responsible for creating/recoverying accounts in a better way at some point, now it's a bit complicated regarding the code structure.

> Looking at the replaceReducer documentation I see the outlined use cases are code-splitting and HMR - neither of which are conventional production runtime use cases like this usage is. It seems excessive to replace the entire root reducer just for adding one new account. 

So we are not replacing the whole root reducer. We are just adding the set of reducers for the new account.

> If the intention here is to populate the state for a new account ID key in byAccountId, then why do we need to replace the entire root reducer? What are the performance and UI implications of doing this?

The intention is not to populate the state, but to inject the reducers dynamically. At the start of the app, we are setting reducers for all accounts that the user has in local storage. But when the user creates/recovers an additional account, we need to add a new set of reducers (account, staking, status, etc.). The result is that every account has its own set of reducers, instead of the current solution where every reducer is holding the state for each account.

> It would be better to iterate on it if we're not yet sure what the final implementation should look like (i.e. TODO: figure out better way to inject reducer).

I mean it is the final implementation of adding reducers dynamically. The next PR will be to figure out a better way than using `saveAccount` method - because it's relevant but I want to split changes into smaller PRs. I mean it's working well as it is, and we are using a similar solution in the project, but we want to get rid of using `store` in `wallet.js` - so the next step will be PR that will base on this one, and will introduce a better way to do it. 

> re: my comment on the top-level reducer, I am picturing this conceptually as a map of account IDs in the redux state whose value is some AccountState type (which, going from the diagram in the ticket, includes tokenBalancesNFT, tokenBalancesFT, etc.). Since you can only have one active account, the active account is tracked by another key at the top level of the state.

So this is how it's implemented now, the only change is that right now we are not holding the shared reducers in the `shared` object, and we don't hold accounts in the `byAccountId` object. But besides that, that is exactly how it's working now. I can easily close all shared reducers in the `shared` object, and all accounts in the `byAccountId` object, like it is on the diagram, I'm just not sure if we need an additional object level. It will look better, the state will be more organized for sure, so we can change that - easy change.

About `activeAccountId` - I'm not sure if we need it, which account is active we know from local storage (it's set in `wallet.accountId`) - so we always know which account is active. I'm not sure if we will ever use `activeAccountId` state from Redux.

Btw my solution is inspired by this comment from Dan Abramov:
https://stackoverflow.com/questions/32968016/how-to-dynamically-load-reducers-for-code-splitting-in-a-redux-application/33044701#33044701
It's a different use case, but it's showing a concept of how to add a new reducer dynamically.

Also this PR:
https://github.com/near/near-wallet/pull/2555
is showing how it will work within an example of `tokens` reducer. It's based on this PR.
We are splitting `tokens` reducer into:
 - `tokens` that is a reducer that every account will have
 - `tokensMetadata ` that is shared.
You can test it by recovering an additional account, you will see that `tokensMetadata` is not replaced or purged.